herddb-core/src/main/java/herddb/core/TableManager.java

```diff
@@ -2080,13 +2102,21 @@ private TableCheckpoint checkpoint(double dirtyThreshold, double fillThreshold,
                     /* Flush the page if it would exceed max page size */
                     final long recordSize = DataPage.estimateEntrySize(record);
                     if (bufferPageSize + recordSize > maxLogicalPageSize) {
-                        createImmutablePage(buffer, bufferPageSize);
+                        /* Do no keep in memory old untouched records */
+                        createImmutablePage(buffer, bufferPageSize, keepFlushedPageInMemory);
+
+                        /* Reset next rebuilt page status */
+                        keepFlushedPageInMemory = false;
+
                         flushedRecords += buffer.size();
                         bufferPageSize = 0;
                         /* Do not clean old buffer! It will used in generated pages to avoid too many copies! */
                         buffer = new HashMap<>(buffer.size());
                     }
 
+                    /* Current rebuilt page will be kept in memory if current page was in memory */
+                    keepFlushedPageInMemory |= currentPageWasInMemory;
```

Comment:
why are we keeping the last page ? 
are we going to modify it and not creating a new "current page" after the checkpoint ?

Reply:
remaining "spare" data will be flushed into newpages if there is enough space. If there is still too "spare" date it will be flushed on a new immutable page after new pages flush. (This behaviour is unchanged, what chaged here is that if a buffer don't contains data from any loaded page the page created from it shouldn't be kept in memory)